core/src/test/scala/org/broadinstitute/dsde/rawls/entities/base/ExpressionEvaluationSupportSpec.scala

```diff
@@ -0,0 +1,127 @@
+      optType.setTypeName(typeName)
+      valueType.setOptionalType(optType)
+      valueType.setTypeName(TypeNameEnum.OPTIONAL)
+      input.setOptional(isOptional)
+    } else if (isArray) {
+      val arrType = new ValueType()
+      arrType.setTypeName(typeName)
+      valueType.setTypeName(TypeNameEnum.ARRAY)
+      valueType.setArrayType(arrType)
+    } else {
+      valueType.setTypeName(typeName)
+    }
+
+    input.setName("inputName") // doesn't matter for these tests
+    input.setValueType(valueType)
+    MethodInput(input, "this.something") // expression value doesn't matter for these tests
+  }
+
+  // helper method to create a unique name for each test
+  def testDescriptor(typeName: TypeNameEnum, isOptional: Boolean, isArray: Boolean) = Seq(
+    if (isOptional) Some("optional") else None,
+    if (isArray) Some("array of") else None,
+    Some(typeName.getValue)
+  ).flatten.mkString(" ")
+
+  // all input types that are not String, nor are top-level optional or array, since those have sub-types
+  val nonStringTypes:Set[TypeNameEnum] = TypeNameEnum.values().toSet diff Set(TypeNameEnum.STRING, TypeNameEnum.OPTIONAL, TypeNameEnum.ARRAY)
+
+  // programmatically create unit tests for optional=true/false, array=true/false, and each type
+  List(true, false) foreach { isOptional =>
+    List(true, false) foreach { isArray =>
+      // these types should return true, i.e. isStringInputType detects a string input
+      List(TypeNameEnum.STRING) foreach { valueType =>
+        it should s"detect string inputs for ${testDescriptor(valueType, isOptional, isArray)}" in {
+          val input = methodInput(valueType, isOptional = isOptional, isArray = isArray)
+          isStringInputType(input) shouldBe true
+        }
+      }
+      // these types should return false, i.e. isStringInputType does not detect a string input
+      nonStringTypes.foreach { valueType =>
+        it should s"not detect a string input for ${testDescriptor(valueType, isOptional, isArray)}" in {
+          val input = methodInput(valueType, isOptional = isOptional, isArray = isArray)
+          isStringInputType(input) shouldBe false
+        }
+      }
+    }
+  }
+
+  behavior of "ExpressionEvaluationSupport.maybeConvertToString"
+
+  // define a map of (input attribute value) -> (expected result from castToString)
+  val stringConversions = Map(
+    AttributeNumber(123) -> AttributeString("123"),
+    AttributeNumber(123.4567) -> AttributeString("123.4567"),
+    AttributeNumber(Double.MaxValue) -> AttributeString("1.7976931348623157E+308"), // should we disable scientific notation??
+    AttributeNumber(Double.MinValue) -> AttributeString("-1.7976931348623157E+308"), // should we disable scientific notation??
+    AttributeNumber(Double.MinPositiveValue) -> AttributeString("4.9E-324"), // should we disable scientific notation??
+    AttributeBoolean(true) -> AttributeString("true"),
+    AttributeBoolean(false) -> AttributeString("false"),
```

Comment:
technically we could cause a case change: User uploads `"False"` -> converted to boolean `false`  -> converted to string `"false"`. This should be fine since workflows likely are converting the string back to a boolean before checking against it

Reply:
This will also depend on how lenient we are during TSV parsing. The previous PR (https://github.com/broadinstitute/firecloud-orchestration/pull/929/files#diff-acb77f680e9d2503b8afd1827a0355015e278ef1313339a6acdfa621788b5f3bR154) allowed both "false" and "False" as boolean strings, so your comment holds, but we could change the TSV parsing when we roll it out again. That said, I hope this is a non-issue!

Reply:
Adding on: as soon as we parse the TSV into a boolean, we've lost its original string case - so users could upload "FaLsE" in their TSV but they'd see "false" in the UI, even if they never submit a workflow